### PR TITLE
Support for 3D registers and layouts in the abstract representation

### LINKED ISSUE #695: Support 3D registers in the abstract representation

### PR DESCRIPTION
**Pull request by HGSilveri**
Closes #695 .

### CODE REVIEW

File: pulser-core/pulser/json/abstract_repr/deserializer.py [hunk 1]
```diff
@@ -589,20 +605,59 @@ def deserialize_abstract_layout(obj_str: str) -> RegisterLayout:
     return _deserialize_layout(json.loads(obj_str))
 
 
-def deserialize_abstract_register(obj_str: str) -> BaseRegister:
+@overload
+def deserialize_abstract_register(
```
Comment (awennersteen) [inline pulser-core/pulser/json/abstract_repr/deserializer.py:609]: All of these overload: pass confuse me.
Is that just the way to do it, or is there a reason for doing it if not required?

File: pulser-core/pulser/json/abstract_repr/deserializer.py [hunk 2]
```diff
@@ -589,20 +605,59 @@ def deserialize_abstract_layout(obj_str: str) -> RegisterLayout:
...
+    obj_str: str, required_dim: Literal[None, 2, 3] = None
+) -> Register | Register3D:
     """Deserialize a register from an abstract JSON object.
 
     Args:
-        obj_str: the JSON string representing the register encoded
+        obj_str: The JSON string representing the register encoded
             in the abstract JSON format.
+        required_dim: If defined, ensures the register is of the
```
Comment (awennersteen) [inline pulser-core/pulser/json/abstract_repr/deserializer.py:635]: Honestly, calling it required_dim to me is a bit strange.

Is just dim, dimension, dimensionality or similar in some way  better?

At the same time this is not something externals will care much about so if the naming makes more sense to you I'm OK with it

File: pulser-core/pulser/json/abstract_repr/deserializer.py (hunk 1 above)
Comment (HGSilveri) [reply pulser-core/pulser/json/abstract_repr/deserializer.py:609]: They are only for `mypy`, they allow it to restrict the return type based on the type (or in this case the actual value) of the input. Without these, I'd have to `cast` the outputs to `Register` or `Register3D` on the `.from_abstract_repr()`

File: pulser-core/pulser/json/abstract_repr/deserializer.py [hunk 3]
```diff
@@ -589,20 +605,59 @@ def deserialize_abstract_layout(obj_str: str) -> RegisterLayout:
...
+            specified dimensionality.
 
     Returns:
         The Register instance.
     """
+    if required_dim not in (None, 2, 3):
+        raise ValueError(
+            "When specified, 'required_dim' must be 2 or 3, "
+            f"not {required_dim!s}."
```
Comment (awennersteen) [inline pulser-core/pulser/json/abstract_repr/deserializer.py:644]: Is the !s a typo, or syntax I'm not familiar with?

File: pulser-core/pulser/json/abstract_repr/deserializer.py (hunk 2 above)
Comment (HGSilveri) [reply pulser-core/pulser/json/abstract_repr/deserializer.py:635]: How about `expected_dim`?

File: pulser-core/pulser/json/abstract_repr/deserializer.py (hunk 1 above)
Comment (awennersteen) [reply pulser-core/pulser/json/abstract_repr/deserializer.py:609]: Thanks!

File: pulser-core/pulser/json/abstract_repr/deserializer.py (hunk 3 above)
Comment (awennersteen) [reply pulser-core/pulser/json/abstract_repr/deserializer.py:644]: Huh

Had no idea. But I see this is not exactly new syntax. Wonder how I've never seen it

File: pulser-core/pulser/json/abstract_repr/deserializer.py (hunk 2 above)
Comment (awennersteen) [reply pulser-core/pulser/json/abstract_repr/deserializer.py:635]: Sounds better to me

**Review by awennersteen — APPROVED** [review]
Lgtm

Thanks!

**Pull request merged**